contrib/babelfishpg_tsql/src/tsql_for/forxml.c

```diff
@@ -0,0 +1,329 @@
+				tsql_for_datetime_format(format_output, val);
+				colval = CStringGetDatum(format_output->data);
+
+				datatype_oid = CSTRINGOID;
+			}
+			/*
+			 * datetimeoffset has two behaviors:
+			 * if offset is 0, just return the datetime with 'Z' at the end
+			 * otherwise, append the offset
+			 */
+			else if (strcmp(typename, "datetimeoffset") == 0)
+			{
+				char *val = SPI_getvalue(tuple, tupdesc, i+1);
+				StringInfo format_output = makeStringInfo();
+				tsql_for_datetimeoffset_format(format_output, val);
+				colval = CStringGetDatum(format_output->data);
+
+				datatype_oid = CSTRINGOID;
+			}
+		}
+		if (!isnull)
+		{
+			appendStringInfo(state, " %s=\"%s\"",
+							 colname,
+							 map_sql_value_to_xml_value(colval, datatype_oid, true));
+		}
+	}
+	appendStringInfoString(state, "/>");
+}
+
+/*
+ * Map an SQL row to an XML element in PATH mode.
+ */
+static void
+tsql_row_to_xml_path(StringInfo state, Datum record, const char* element_name, bool binary_base64)
+{
+	HeapTupleHeader td;
+	Oid				tupType;
+	int32			tupTypmod;
+	TupleDesc		tupdesc;
+	HeapTupleData 	tmptup;
+	HeapTuple		tuple;
+	bool 			allnull = true;
+
+	td = DatumGetHeapTupleHeader(record);
+
+	/* Extract rowtype info and find a tupdesc */
+	tupType = HeapTupleHeaderGetTypeId(td);
+	tupTypmod = HeapTupleHeaderGetTypMod(td);
+	tupdesc = lookup_rowtype_tupdesc(tupType, tupTypmod);
+
+	/* Build a temporary HeapTuple control structure */
+	tmptup.t_len = HeapTupleHeaderGetDatumLength(td);
+	tmptup.t_data = td;
+	tuple = &tmptup;
+
+	/* each tuple is either contained in a "row" tag, or standalone if the element_name is an empty string */
+	if (element_name[0] != '\0') // if "''" is the input path, ignore it per SQL Server behavior
+		appendStringInfo(state, "<%s>", element_name);
+
+	/* process the tuple into tags */
+	for (int i = 0; i < tupdesc->natts; i++)
+	{
+		char	*colname;
+		Datum	colval;
+		bool	isnull;
+		Oid 	datatype_oid;
+		Oid 	nspoid;
+		Oid 	tsql_datatype_oid;
+		char	*typename;
+		Form_pg_attribute att = TupleDescAttr(tupdesc, i);
+
+		if (att->attisdropped)
+			continue;
+
+		colname = map_sql_identifier_to_xml_name(NameStr(att->attname), true, false);
+		colval = heap_getattr(tuple, i + 1, tupdesc, &isnull);
+		datatype_oid = att->atttypid;
+
+		/* 
+		 * Below is a workaround for is_tsql_x_datatype() which does not work as expected.
+		 * We compare the datatype oid of the columns with the tsql_datatype_oid and
+		 * then specially handle some TSQL-specific datatypes.
+		 */
+		typename = SPI_gettype(tupdesc, i+1);
+		nspoid = get_namespace_oid("sys", true);
+		Assert(nspoid != InvalidOid);
+
+		tsql_datatype_oid = GetSysCacheOid2(TYPENAMENSP, Anum_pg_type_oid, CStringGetDatum(typename), ObjectIdGetDatum(nspoid));
+	
+		/*
+		 * tsql_datatype_oid can be different from datatype_oid when there are datatypes in different namespaces
+		 * but with the same name. Examples: bigint, int, etc.
+		 */
+		if (tsql_datatype_oid == datatype_oid)
```

Comment:
This check logic can be shared at least between `tsql_row_to_xml_path()` and `tsql_row_to_xml_raw()`.

Reply:
Extracted the shared code into a separate function.